File: app/src/main/java/com/orgzly/android/prefs/IntegerPreference.java
package com.orgzly.android.prefs;

import android.content.Context;
import android.content.res.TypedArray;
import android.preference.EditTextPreference;
import android.util.AttributeSet;

import com.orgzly.R;

public class IntegerPreference extends EditTextPreference {
    private int min = Integer.MIN_VALUE;
    private int max = Integer.MAX_VALUE;

    public IntegerPreference(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        parseAttrs(attrs);
    }

    public IntegerPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        parseAttrs(attrs);
    }

    private void parseAttrs(AttributeSet attrs) {
        if (attrs != null) {
            TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.IntegerRange);

            min = typedArray.getInt(R.styleable.IntegerRange_min, Integer.MIN_VALUE);
            max = typedArray.getInt(R.styleable.IntegerRange_max, Integer.MAX_VALUE);

            typedArray.recycle();
        }
    }

    @Override
    protected void onDialogClosed(boolean positiveResult) {
        String value = getEditText().getText().toString();
        super.onDialogClosed(validateIntegerValue(value) != null);
    }


    @Override
    public void setText(String text) {
        text = validateIntegerValue(text);

        if (text != null) {
            super.setText(text);

            setSummary(text);
        }
    }

    private String validateIntegerValue(String s) {
        if (s != null && s.length() > 0) {
            try {
                int n = Integer.parseInt(s);
                if (min <= n && n <= max) {
                    return String.valueOf(n);
                }
            } catch (NumberFormatException e) {
                e.printStackTrace();
            }
        }

        return null;
    }
}

File: app/src/main/java/com/orgzly/android/prefs/IntegerPreference.kt
package com.orgzly.android.prefs

import android.content.Context
import android.util.AttributeSet
import androidx.preference.EditTextPreference
import com.orgzly.R
import com.orgzly.android.ui.util.styledAttributes

class IntegerPreference : EditTextPreference {
    private var min = Integer.MIN_VALUE
    private var max = Integer.MAX_VALUE

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        parseAttrs(attrs)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        parseAttrs(attrs)
    }

    private fun parseAttrs(attrs: AttributeSet?) {
        if (attrs != null) {
            context.styledAttributes(attrs, R.styleable.IntegerRange) { typedArray ->
                min = typedArray.getInt(R.styleable.IntegerRange_min, Integer.MIN_VALUE)
                max = typedArray.getInt(R.styleable.IntegerRange_max, Integer.MAX_VALUE)
            }
        }
    }

    override fun setText(text: String) {
        validateIntegerValue(text)?.let {
            super.setText(it)
            summary = it
        }
    }

    private fun validateIntegerValue(s: String): String? {
        if (s.isNotEmpty()) {
            try {
                val n = Integer.parseInt(s)
                if (n in min..max) {
                    return n.toString()
                }
            } catch (e: NumberFormatException) {
                e.printStackTrace()
            }
        }

        return null
    }
}
